File: benchmarks/src/jmh/java/jetbrains/exodus/benchmark/dataStructures/persistent/JMHPersistentLong23TreeBenchmark.java
/**
 * Copyright 2010 - 2017 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.benchmark.dataStructures.persistent;

import jetbrains.exodus.core.dataStructures.persistent.PersistentLong23TreeMap;
import org.openjdk.jmh.annotations.*;

import java.util.TreeMap;
import java.util.concurrent.TimeUnit;

@State(Scope.Thread)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
public class JMHPersistentLong23TreeBenchmark {

    private static final int MAP_SIZE = 100000;

    private final PersistentLong23TreeMap<Object> tree = new PersistentLong23TreeMap();
    private final TreeMap<Long, Object> juTree = new TreeMap<>();
    private final Object value = new Object();
    private long existingKey = 0;
    private long missingKey = MAP_SIZE;

    @Setup
    public void prepare() {
        final PersistentLong23TreeMap<Object>.MutableMap mutableMap = tree.beginWrite();
        for (int i = 0; i < MAP_SIZE; ++i) {
            // the keys are even
            mutableMap.put((long) (i * 2), value);
            juTree.put((long) (i * 2), value);
        }
        mutableMap.endWrite();
    }

    @Setup(Level.Invocation)
    public void prepareKeys() {
        // the even key exists in the map, the odd one doesn't
        existingKey = (long) ((Math.random() * MAP_SIZE) * 2);
        missingKey = existingKey + 1;
    }

    @Benchmark
    @Warmup(iterations = 4, time = 1)
    @Measurement(iterations = 6, time = 1)
    @Fork(5)
    public Object getExisting() {
        return tree.beginRead().get(existingKey);
    }

    @Benchmark
    @Warmup(iterations = 4, time = 1)
    @Measurement(iterations = 6, time = 1)
    @Fork(5)
    public Object getMissing() {
        return tree.beginRead().get(missingKey);
    }

    @Benchmark
    @Warmup(iterations = 4, time = 1)
    @Measurement(iterations = 6, time = 1)
    @Fork(5)
    public Object treeMapGetExisting() {
        return juTree.get(existingKey);
    }

    @Benchmark
    @Warmup(iterations = 4, time = 1)
    @Measurement(iterations = 6, time = 1)
    @Fork(5)
    public Object treeMapGetMissing() {
        return juTree.get(missingKey);
    }
}

File: benchmarks/src/jmh/kotlin/jetbrains/exodus/benchmark/dataStructures/persistent/JMHPersistentLong23TreeBenchmark.kt
/**
 * Copyright 2010 - 2017 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.benchmark.dataStructures.persistent

import jetbrains.exodus.core.dataStructures.persistent.PersistentLong23TreeMap
import jetbrains.exodus.core.dataStructures.persistent.read
import jetbrains.exodus.core.dataStructures.persistent.writeFinally
import org.openjdk.jmh.annotations.*
import org.openjdk.jmh.infra.Blackhole
import java.util.*
import java.util.concurrent.TimeUnit

@State(Scope.Thread)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
open class JMHPersistentLong23TreeBenchmark {

    companion object {
        const val MAP_SIZE = 100000
    }

    private val tree = PersistentLong23TreeMap<Any>()
    private val juTree = TreeMap<Long, Any>()
    private val value = Object()
    private var existingKey: Long = 0
    private var missingKey: Long = MAP_SIZE.toLong()

    @Setup
    fun prepare() {
        tree.writeFinally {
            for (i in 0..MAP_SIZE - 1) {
                // the keys are even
                put(i.toLong() * 2, value)
                juTree[i.toLong() * 2] = value
            }
        }
    }

    @Setup(Level.Invocation)
    fun prepareKeys() {
        // the even key exists in the map, the odd one doesn't
        existingKey = (Math.random() * MAP_SIZE).toLong() * 2
        missingKey = existingKey + 1
    }

    @Benchmark
    @Warmup(iterations = 4, time = 1)
    @Measurement(iterations = 6, time = 1)
    @Fork(5)
    fun getExisting(bh: Blackhole) = bh.consume(tree.read { get(existingKey) })

    @Benchmark
    @Warmup(iterations = 4, time = 1)
    @Measurement(iterations = 6, time = 1)
    @Fork(5)
    fun getMissing(bh: Blackhole) = bh.consume(tree.read { get(missingKey) })

    @Benchmark
    @Warmup(iterations = 4, time = 1)
    @Measurement(iterations = 6, time = 1)
    @Fork(5)
    fun treeMapGetExisting(bh: Blackhole) = bh.consume(juTree[existingKey])

    @Benchmark
    @Warmup(iterations = 4, time = 1)
    @Measurement(iterations = 6, time = 1)
    @Fork(5)
    fun treeMapGetMissing(bh: Blackhole) = bh.consume(juTree[missingKey])
}